### DISCUSSION TITLE
Align earthaccess with STAC

### DISCUSSION THREAD

**Discussion by betolink** [earthaccess-dev/earthaccess#221] [Ideas]
Based on what the GIS/Pangeo communities are doing, I think earthaccess could be more useful if it was aligned with the STAC specification. An idea would be to support both CMR and CMR-STAC access patterns, @scottyhq suggested something like this already in #167 

```python
import earthaccess 

# we could even use NOAA's catalog or any other STAC endpoint, the default will still be NASA CMR (no STAC)
earthaccess.configure({
 "search": "https://cmr.earthdata.nasa.gov/cloudstac/",
 "protocol": "STAC"
})

# when we use STAC, earthaccess will be wrapping pystac_client
items = earthaccess.search_data(
    collections = ['ATL06', 'ATL08', '3IMERGHH'],
    intersects=dict(type="Point", coordinates=[-65.78, 45.79]),
    datetime = '2018/2024',
)

# earthacces will have to know how to deal with STAC results
earthaccess.download(items, "./data")

```

It would be interesting to explore [`intake-stac`](https://github.com/intake/intake-stac/blob/main/examples/intake-cmr-stac.ipynb) to stream the results and earthaccess will only provide the authenticated fsspec sessions.

This integration will make earthaccess more flexible but there are some things that we need to verify before trying to implement this:

* Will CMR-STAC results be consistent with the results provided by CMR?
* There is no top level NASA catalog, a user needs to know that ATL is in the NSIDC catalog and 3IMERGHH in GESDISC, can there be a way to avoid this so the users don't need to know where each dataset is?
* Can we interchangeable use `DOI` / `short_name` to search for data collections?
* What is the difference between say
    * https://cmr.earthdata.nasa.gov/stac/LPCLOUD vs https://cmr.earthdata.nasa.gov/cloudstac/LPCLOUD

**Comment by BriannaLind** [comment]
That makes a lot of sense to me!  I have no idea of how to go about implementing that though!

**Comment by betolink** [comment]
Some answers from the CMR-STAC team

---

1. Results on CLOUDSTAC will have the cloudHosted attribute for their parent collections
    * `/cloudstac/<provider>` can be considered a subset of `/stac/<provider>`
    * in the case of LPCLOUD, if there are any collections NOT marked as cloudHosted, they should not show up in /cloudstac/LPCLOUD, but would be accessible in /stac/LPCLOUD

2. CMR-STAC is a STAC compliant wrapper for queries into CMR. There may still be some minor inconsistencies due to currently unsupported query mechanisms
    * e.g. a POST search `{"query": {"eo:cloud_cover": {"gte": 50 } }`  may yield unexpected results as  CMR only supports a min and max value, but not other comparison operators
    * in intersects queries with polygons containing holes, the holes are not honored presently and you would need to use Shapefile searches against CMR directly

3. Searching across all providers is something that has been discussed, but it not currently supported. Knowing there is a demand for it will help gauge the priority for adding that.
---

Based on this **I think** aligning earthaccess with CMR-STAC is doable and will bring great flexibility and re-usability to existing workflows.

**Reply by abbottry** [reply]
1. CMR STAC and CLOUDSTAC should be merged and a parameter should allow filtering for `cloudHosted` -- we're working towards that. 
2. You are correct
3. This is also correct, its difficult to allow searching across ~2 Billion records so we're looking into ways to allow this for uses like STAC where its more necessary.

**Comment by scottyhq** [comment]
Thanks for documenting all these ideas @betolink ! 

> [@scottyhq](https://github.com/scottyhq) suggested something like this already in [#167](https://github.com/nsidc/earthaccess/issues/167)

Yes, my suggestion was to support the same search *keywords and values* as the STAC API (e.g. `intersects=dict(type="Point", coordinates=[lon, lat]`) and translate those to whatever CMR needs if necessary, rather than supporting endpoints other than CMR 


> CMR-STAC is a STAC compliant wrapper for queries into CMR. There may still be some minor inconsistencies due to currently unsupported query mechanisms

Based on this, I can see why the above suggestion might be hard!

In practice I find this lack of full compliance super difficult as a user. I’ve tried to use CMR-STAC multiple times over the last couple years but unfortunately each time find that it falls short and I stop using it: 

 1. It’s really slow compared to directly searching CMR and getting JSON outputs
 2. The API inconsistencies manifest as issues when people come from other STAC API endpoints like USGS Collection2 or Planetary Computer (see https://github.com/nasa/cmr-stac/issues)
 3. STAC is most useful with a bunch of metadata *properties* per item (and extensions like `eo`,`proj`,`sar`,etc) but CMR-STAC only returns the basic datetime and polygon for most items

So I wonder instead of the proxy approach, what is preventing hosting STAC metadata directly in CMR (at least just for the new cloudHosted datasets)? 

> It would be interesting to explore [intake-stac](https://github.com/intake/intake-stac/blob/main/examples/intake-cmr-stac.ipynb) to stream the results and earthaccess will only provide the authenticated fsspec sessions.

Providing fsspec objects is nice for consumption by downstream tools. I'll caution that intake-stac worked well for the case of a arbitrary single file (CSV, HDF, TIF, etc) to give back a python object, but because searches can return pretty complicated *collections* of data, the process of returning a composite data object in an efficient and accurate way is pretty hard. You start getting into fsspec cache settings, and other settings based on the tool that ultimately does I/O (e.g. rasterio/GDAL). This is where development of intake-stac basically stalled. If you restrict the possible data formats it does become much more tractable. For example just dealing with COGs is complicated enough with [stackstac](https://stackstac.readthedocs.io/) and [odc-stac](https://odc-stac.readthedocs.io/) :)

**Reply by scottyhq** [reply]
Hmmm, I just noticed the CMR API lists STAC as a supported return type https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#stac , is that actually using cmr-stac?

**Reply by abbottry** [reply]
Hi @scottyhq we worked on a feature a bit ago to support this but ultimately the format isn't fully compliant, for (mostly) that reason we're looking at removing that support in favor of a more concerted effort to improve CMR Stac.

**Comment by cboettig** [comment]
Hi @abbottry -- is the a decent place to report other issues about the EarthData STAC?   I wanted to note that in addition to the issues mentioned above, the static JSON versions could be improved, e.g. for better consistency with stac-browser.

Note that currently stac-browser does not recognize either the pagination or the stac API ([rendered example](https://radiantearth.github.io/stac-browser/#/external/cmr.earthdata.nasa.gov/stac/POCLOUD?page=3&.language=en)) makes it appear like there are only 10 entries.

As NASA is already providing both static JSON and dynamic API endpoints, I understand the API should be mentioned in the JSON using the `conformance` tag:

```
{
    "rel": "conformance",
      "type": "application/json",
      "title": "STAC/WFS3 conformance classes implemented by this server",
      "href": "https://planetarycomputer.microsoft.com/api/stac/v1/conformance"
    }
```

Examples of this can be seen in, e.g. on [planetary computer stac](https://radiantearth.github.io/stac-browser/#/external/planetarycomputer.microsoft.com/api/stac/v1/collections/landsat-c2-l2) (note the API box at top and the added search abilities that come with this; see [source json](https://planetarycomputer.microsoft.com/api/stac/v1/collections/landsat-c2-l2) for comparison).

**Comment by scottyhq** [comment]
It's been a while and I thought I'd revive this discussion. I'm still a bit confused by 1. NASA CMR already directly supporting STAC as an [output type](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#stac) versus 2. The [NASA-CMR STAC Proxy](https://github.com/nasa/cmr-stac) effort.  

If the cmr-stac proxy is the future of development that would be good to know - as then `earthaccess` could use `pystac_client` as a dependency as @betolink describes above!

However, another possible approach is to *just expose the STAC output type from CMR*. I just opened an issue about that here https://github.com/nasa/python_cmr/issues/81 . This would make it straightforward to take advantage of other great tools like `pystac` and `geopandas` for spatial operations, map exploration (https://github.com/nsidc/earthaccess/issues/305) etc...

The example below would require a small change to `python_cmr`, but since earthaccess uses python_cmr should be easy to integrate?...
```python
import geopandas as gpd
from cmr import GranuleQuery
import json

api = GranuleQuery() 
search = api.parameters(
    point=(-105.78, 35.79),
    temporal=('2021-02-01','2021-03-01'),
    collection_concept_id='C2021957657-LPCLOUD'  # Required for STAC search
)
items = search.format("stac").get()
features = json.loads(items[0]) # python_cmr returns list with string-wrapped dict
gf = gpd.GeoDataFrame.from_features(features, crs='EPSG:4326')
```

**Reply by chuckwondo** [reply]
Hi @scottyhq. I just replied to your python_cmr issue.

Regarding your example above, I would recommend using `json.loads` rather than `ast.literal_eval`, not only because most folks will be more familiar with it, but also because the strings are in fact JSON strings.

We also need to handle things a bit differently when paging is involved, but perhaps it would be better for me to go into detail in the issue you opened: https://github.com/nasa/python_cmr/issues/81

**Reply by jhkennedy** [reply]
> I'm still a bit confused by 1. NASA CMR already directly supporting STAC as an [output type](https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html#stac) versus 2. The [NASA-CMR STAC Proxy](https://github.com/nasa/cmr-stac) effort.

From the comment above, https://github.com/nsidc/earthaccess/discussions/221#discussioncomment-7534548, I think this implies that the STAC output type is (or will be) deprecated, and future efforts will revolve around improving the STAC proxy

@abbottry or @doug-newman-nasa can you confirm?

**Reply by doug-newman-nasa** [reply]
Any effort to use STAC with CMR should use the CMR STAC API. I'm loathed to use the word 'proxy' when describing it, this just adds to the confusion. :)

**Reply by scottyhq** [reply]
thanks for the insight @doug-newman-nasa! 

> I'm loathed to use the word 'proxy' when describing it, this just adds to the confusion. :)

Could you clarify why? Perhaps the readme needs updating, as it says up front "CMR-STAC acts as a proxy between the CMR repository and STAC API queries." https://github.com/nasa/cmr-stac?tab=readme-ov-file#cmr-stac

**Reply by doug-newman-nasa** [reply]
The README does need updating. Why use the word proxy? It is CMR's STAC API. It's a design detail.

**Reply by jhkennedy** [reply]
@doug-newman-nasa, thanks for weighing in! I agree, CMR STAC API is more appropriate as the proxy part is largely irrelevant to end-users.

**Comment by cboettig** [comment]
Perhaps orthogonal to the issue of a stac API, but this sounds really compelling to me: https://cloudnativegeo.org/blog/2024/08/introduction-to-stac-geoparquet/

Specifically, it's really nice to be able to see the entire catalog metadata _as data_ like this, rather than having to pound the heck out of the API...
